### PR TITLE
WIP dynamic conv pr

### PR DESCRIPTION
**Pull request by HDCharles**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #67982

Summary: TODO test using dynamic_conv with the quantization
workflow note: found a bug with quantized implementation of ConvTranspose3d
which needed to be corrected
Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

### CODE REVIEW

File: aten/src/ATen/native/quantized/cpu/qconv_dynamic.cpp [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+#include <torch/library.h>
+
+#ifdef USE_FBGEMM
+
+template <int kSpatialDim>
+at::Tensor PackedConvWeight<kSpatialDim>::apply_dynamic(
+    const at::Tensor& input,
+    bool reduce_range) {
+  /// is this needed?
+  auto input_contig = input.contiguous();
```
Comment (kimishpatel) [inline aten/src/ATen/native/quantized/cpu/qconv_dynamic.cpp:23]: This should be channels last contiguous. fbgemm expects input to be in channels last format. Check the equivalent qconv implementation.

File: aten/src/ATen/native/quantized/cpu/qconv_dynamic.cpp [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+        /*force_scale_power_of_two=*/false,
+        /*reduce_range=*/reduce_range);
+
+    // Quantize input
+    at::Tensor q_input = at::quantize_per_tensor(
+        input, q_params.scale, q_params.zero_point, c10::kQUInt8);
+
+    at::Tensor out = apply_impl<false>(q_input, q_params.scale, q_params.zero_point);
+
+    return at::dequantize(out);
```
Comment (kimishpatel) [inline aten/src/ATen/native/quantized/cpu/qconv_dynamic.cpp:110]: Put a comment here to the effect that we can have optimized kernel that produces fp32 output. This also holds for fbgemm.

File: aten/src/ATen/native/quantized/cpu/qconv_dynamic.cpp (hunk 1 above)
Comment (kimishpatel) [reply aten/src/ATen/native/quantized/cpu/qconv_dynamic.cpp:23]: Actually if we want this to work for transposed conv as well, I think it is better to just not call `contiguous` here. For doing per tensor quantization I dont think it matters.

File: aten/src/ATen/native/quantized/cpu/qconv_dynamic.cpp [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+  m.impl(TORCH_SELECTIVE_NAME("quantized::conv_transpose3d_dynamic"),
+  TORCH_FN(QConvDynamicInt8<3>::run));
+}
+
+/* is this needed?
+TORCH_LIBRARY_IMPL(_quantized, CPU, m) {
+  m.impl(TORCH_SELECTIVE_NAME("_quantized::linear_dynamic"),
+TORCH_FN(QLinearDynamicInt8<false>::run));
+}*/
+
```
Comment (kimishpatel) [inline aten/src/ATen/native/quantized/cpu/qconv_dynamic.cpp:181]: Remove.

File: aten/src/ATen/native/quantized/library.cpp [hunk 4]
```diff
@@ -99,6 +104,10 @@ TORCH_LIBRARY(quantized, m) {
   m.def(TORCH_SELECTIVE_SCHEMA("quantized::conv_transpose1d(Tensor qx, __torch__.torch.classes.quantized.Conv2dPackedParamsBase packed_weight, float output_scale, int output_zero_point) -> Tensor"));
   m.def(TORCH_SELECTIVE_SCHEMA("quantized::conv_transpose2d(Tensor qx, __torch__.torch.classes.quantized.Conv2dPackedParamsBase packed_weight, float output_scale, int output_zero_point) -> Tensor"));
   m.def(TORCH_SELECTIVE_SCHEMA("quantized::conv_transpose3d(Tensor qx, __torch__.torch.classes.quantized.Conv3dPackedParamsBase packed_weight, float output_scale, int output_zero_point) -> Tensor"));
+  // do I need to add these 3 to _quantized?
```
Comment (kimishpatel) [inline aten/src/ATen/native/quantized/library.cpp:107]: What do you mean by `_quantized`?

File: test/quantization/core/test_quantized_op.py [hunk 5]
```diff
@@ -3661,6 +3661,26 @@ def test_embedding_bag_2d_indices(self):
         torch.testing.assert_close(result, qresult, atol=0.05, rtol=1e-3)
 
 
+class TestDynamicQuantizedConv(TestCase):
+    def test_qconv(self):
+        X_fp32 = torch.randn(3, 3, 4, 4)
+        reduce_range = True
+        s, z = _calculate_dynamic_qparams(X_fp32, torch.quint8, reduce_range)
+        X_q = torch.quantize_per_tensor(X_fp32, s, z, torch.quint8)
```
Comment (kimishpatel) [inline test/quantization/core/test_quantized_op.py:3669]: I think general practice has been to quantize the tensor and then dequantize to get `X_fp32` due to loss in precision if you directly operator on original X_fp32. Check other tests in the file.

Comment (kimishpatel) [reply test/quantization/core/test_quantized_op.py:3669]: Also maybe have more test cases generated similar to other tests?

File: torch/ao/quantization/quantization_mappings.py [hunk 6]
```diff
@@ -42,6 +42,7 @@
     nn.Conv3d: nnq.Conv3d,
     nn.ConvTranspose1d: nnq.ConvTranspose1d,
     nn.ConvTranspose2d: nnq.ConvTranspose2d,
+    nn.ConvTranspose3d: nnq.ConvTranspose3d,  # @reviewer, any reason this wasn't here before?
```
Comment (kimishpatel) [inline torch/ao/quantization/quantization_mappings.py:45]: No idea.

File: torch/nn/quantized/dynamic/modules/conv.py [hunk 7]
```diff
@@ -0,0 +1,474 @@
+# coding=utf-8
+r"""Quantized convolution modules."""
+
+from typing import Optional, List, TypeVar
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+# import torch.nn.intrinsic as nni
+# import torch.nn.intrinsic.qat as nniqat
```
Comment (kimishpatel) [inline torch/nn/quantized/dynamic/modules/conv.py:10]: Remove

File: torch/nn/quantized/dynamic/modules/conv.py [hunk 8]
```diff
@@ -0,0 +1,474 @@
...
+    def __init__(self, in_channels, out_channels, kernel_size, stride=1,
+                 padding=0, dilation=1, groups=1, bias=True,
+                 padding_mode='zeros', device=None, dtype=None):
+        factory_kwargs = {'device': device, 'dtype': dtype}
+        kernel_size = _pair(kernel_size)
+        stride = _pair(stride)
+        padding = _pair(padding)
+        dilation = _pair(dilation)
+        # Subclasses of _ConvNd need to call _init rather than __init__. See
+        # discussion on PR #49702
```
Comment (kimishpatel) [inline torch/nn/quantized/dynamic/modules/conv.py:180]: outdated comment? Probably from copying conv module.

File: torch/nn/quantized/dynamic/modules/conv.py [hunk 9]
```diff
@@ -0,0 +1,474 @@
...
+        return 'DynamicQuantizedConv2d'
+
+    def forward(self, input, reduce_range=True):
+        # Temporarily using len(shape) instead of ndim due to JIT issue
+        # https://github.com/pytorch/pytorch/issues/23890
+        if len(input.shape) != 4:
+            raise ValueError("Input shape must be `(N, C, H, W)`!")
+        if self.padding_mode != 'zeros':
+            _reversed_padding_repeated_twice = _reverse_repeat_padding(self.padding)
+            input = F.pad(input, _reversed_padding_repeated_twice,
```
Comment (kimishpatel) [inline torch/nn/quantized/dynamic/modules/conv.py:195]: Is this to pad with specific value? If so for quantized conv does this value need to be quantized as well?

File: torch/nn/quantized/dynamic/modules/conv.py [hunk 10]
```diff
@@ -0,0 +1,474 @@
...
+        return ops.quantized.conv3d_dynamic(
+            input, self._packed_params, reduce_range=reduce_range)
+
+class ConvTranspose1d(nnq.ConvTranspose1d):
+    r"""Applies a 1D transposed convolution operator over an input image
+    composed of several input planes.
+    For details on input arguments, parameters, and implementation see
+    :class:`~torch.nn.ConvTranspose1d`.
+
+    .. note:: Currently only the QNNPACK engine is implemented.
```
Comment (kimishpatel) [inline torch/nn/quantized/dynamic/modules/conv.py:279]: This may not be true. But again probably here from copying.

File: torch/nn/quantized/modules/conv.py [hunk 11]
```diff
@@ -195,7 +195,7 @@ def get_qconv(cls, mod, activation_post_process, weight_post_process=None):
         if weight_post_process is None:
             weight_post_process = mod.qconfig.weight()
         weight_post_process(mod.weight)
-        act_scale, act_zp = activation_post_process.calculate_qparams()
```
Comment (kimishpatel) [inline torch/nn/quantized/modules/conv.py:198]: Why this line is removed?

File: torch/nn/quantized/modules/conv.py [hunk 12]
```diff
@@ -204,6 +204,11 @@ def get_qconv(cls, mod, activation_post_process, weight_post_process=None):
                     mod.stride, mod.padding, mod.dilation, mod.groups,
                     mod.bias is not None, mod.padding_mode)
         qconv.set_weight_bias(qweight, mod.bias)
+        if isinstance(activation_post_process, torch.ao.quantization.PlaceholderObserver):
```
Comment (kimishpatel) [inline torch/nn/quantized/modules/conv.py:207]: I see. It is added here.

File: torch/nn/quantized/modules/conv.py (hunk 11 above)
Comment (kimishpatel) [reply torch/nn/quantized/modules/conv.py:198]: Also, if you leave it as is, what happens? Also do you not need from_float in dynamic conv module?

File: torch/nn/quantized/modules/conv.py (hunk 12 above)
Comment (kimishpatel) [reply torch/nn/quantized/modules/conv.py:207]: And in what instance this condition is true?

**Review by kimishpatel — CHANGES_REQUESTED** [review]

File: test/quantization/core/test_quantized_op.py (hunk 5 above)
Comment (HDCharles) [reply test/quantization/core/test_quantized_op.py:3669]: okay, will do
